filament/backend/src/metal/MetalShaderCompiler.h

```diff
@@ -41,36 +43,74 @@ class MetalShaderCompiler {
 public:
     class MetalFunctionBundle {
     public:
-        MetalFunctionBundle() = default;
-        MetalFunctionBundle(id<MTLFunction> fragment, id<MTLFunction> vertex)
-            : functions{fragment, vertex} {
+        using Raster = std::tuple<id<MTLFunction>, id<MTLFunction>>;
+        using Compute = id<MTLFunction>;
+        using Error = NSString*;
+        struct None {};
+
+        MetalFunctionBundle() : mPrograms{None{}} {}
+
+        explicit operator bool() const { return isValid(); }
+
+        bool isValid() const noexcept {
+            return std::holds_alternative<Raster>(mPrograms) ||
+                std::holds_alternative<Compute>(mPrograms);
+        }
+
+        void validate() const {
+            if (UTILS_LIKELY(!std::holds_alternative<Error>(mPrograms))) {
+                return;
+            }
+            NSString* errorMessage = std::get<Error>(mPrograms);
+            NSString* reason =
+                    [NSString stringWithFormat:
+                            @"Attempting to draw with an id<MTLFunction> that "
+                            @"failed to compile. Compiler error: %@", errorMessage];
+            NSException* compilationFailureException =
+                    [NSException exceptionWithName:@"MetalCompilationFailure"
+                                            reason:reason
+                                          userInfo:nil];
+            [compilationFailureException raise];
+        }
```

Comment:
The "slow" path of vlaidate() should be in the cpp file, so it's not inlined. It should probably be marked as UTILS_NO_INLINE too.